src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs

```diff
@@ -555,13 +567,14 @@ public void ForEach(Action<T> action)
 
             int version = _version;
 
-            for (int i = 0; i < _size; i++)
+            // Use Span to avoid additional bounds checks from using List
+            foreach (T item in new ReadOnlySpan<T>(_items, 0, _size))
```

Comment:
My assumption is this is worthwhile (or, at least, not harmful) for pretty much any size list, since we're avoiding a couple of checks per iteration and only paying a couple of additional checks when creating the span.  But could you run some measurements to validate that?

Reply:
For empty/very small arrays this may actually be a regression. Sadly the span constructor has an explicitly null check that cannot be removed and bloats the code.